### PR TITLE
Delegation implementation: step 1

### PR DESCRIPTION
**Pull request by Bryanskiy**
See https://github.com/rust-lang/rust/issues/118212 for more details.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_resolve/src/late.rs [hunk 1]
```diff
@@ -1331,6 +1343,10 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
         opt_ns: Option<Namespace>, // `None` indicates a module path in import
         finalize: Option<Finalize>,
     ) -> PathResult<'a> {
+        if self.in_delegation_item && path.len() == 1 && path[0].ident.name == kw::SelfLower {
+            return PathResult::NonModule(def::PartialRes::new(def::Res::Err));
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late.rs:1347]: This is hacky and subtly incorrect.

I would suggest creating a new `Res::Local(id)` binding for `self` when entering the delegation item body.
Then all its uses in the body will automatically be resolved correctly.
`fresh_binding` can be used for that (see its other uses for examples).
`in_delegation_item` will no longer be necessary in that case.

Node id can be taken from the delegation item.
The a pass similar to `SelfResolver` will be required to update it to the first parameter's id.
Or you can reserve a `NodeId` for the future first parameter in advance.
Not sure what would be simpler (there may be issues if the id is reserved, but the first parameter won't actually exist).

File: compiler/rustc_resolve/src/late.rs [hunk 2]
```diff
@@ -3040,6 +3064,21 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
...
+                    item.id,
+                    item.ident,
+                    &item.kind,
+                    ValueNS,
+                    item.span,
+                    seen_trait_items,
+                    |i, s, c| MethodNotMemberOfTrait(i, s, c),
+                );
+
+                // do not emit same errors twice during path resolution
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late.rs:3079]: They are not the same errors.
One error says that the path is unresolved, and another says that there's no such item in the trait, both may be reported independently.
I wouldn't bother deduplicating them.

File: compiler/rustc_resolve/src/late.rs [hunk 3]
```diff
@@ -3146,6 +3189,23 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
         })
     }
 
+    fn resolve_delegation(&mut self, delegation: &'ast Delegation, with_path: bool) {
+        self.in_delegation_item = true;
+
+        let (qself, path) = &delegation.path;
+        if let Some(qself) = &delegation.path.0 {
+            self.visit_ty(&qself.ty);
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late.rs:3197]: Need to do full recursive visiting like this (copypasted from visit.rs):
```rust
            if let Some(qself) = maybe_qself {
                visitor.visit_ty(&qself.ty);
            }
            visitor.visit_path(path, expression.id)
```
The path may have generic arguments, etc, which also needs to be resolved.

Also, recursive visiting typically happens after the current node visiting, `smart_resolve_path` in this case.

File: compiler/rustc_resolve/src/late.rs [hunk 4]
```diff
@@ -4572,7 +4632,8 @@ impl<'ast> Visitor<'ast> for LifetimeCountVisitor<'_, '_, '_> {
             | ItemKind::ExternCrate(..)
             | ItemKind::MacroDef(..)
             | ItemKind::GlobalAsm(..)
-            | ItemKind::MacCall(..) => {}
+            | ItemKind::MacCall(..)
+            | ItemKind::Delegation(..) => {}
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late.rs:4636]: This is a bit subtle.
Delegated functions have lifetimes, their count is not necessarily zero.
But skipping the delegation items here doesn't mean that the count will be considered zero, it means there will be a panic when retrieving the count, but for delegation items we are never actually retrieving that count in practice.
This case needs a comment.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 5]
```diff
@@ -221,6 +221,18 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
             let mut span_label = None;
             let item_ident = path.last().unwrap().ident;
             let item_span = item_ident.span;
+            if let PathSource::Delegation = source {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:224]: Not sure why a special error is required here.
What happens if it is removed?

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 6]
```diff
@@ -435,7 +447,13 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
             code,
         );
 
+        // FIXME: suggestions do not supported yet for delegation item
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:450]: Not sure why a special error is required here.
What happens if it is removed?

File: compiler/rustc_middle/src/ty/mod.rs [hunk 7]
```diff
@@ -201,6 +201,10 @@ pub struct ResolverAstLowering {
 
     /// Lints that were emitted by the resolver and early lints.
     pub lint_buffer: Steal<LintBuffer>,
+
+    /// Information about functions signatures for delegation items expansion
+    pub has_self: LocalDefIdSet,
+    pub fns_arguments_count: FxHashMap<DefId, usize>,
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/ty/mod.rs:207]: Please, make recursive delegation (delegating to another delegation item) an error.
These tables are not filled for delegation items, so recursive delegation items won't use correct values here.

They can be made to work by making `has_self` and `fn_argument_count` queries so they could be called recursively, but it's probably a task for a different PR.

File: compiler/rustc_ast_lowering/src/item.rs [hunk 8]
```diff
@@ -427,6 +427,16 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 let macro_def = self.arena.alloc(ast::MacroDef { body, macro_rules: *macro_rules });
                 hir::ItemKind::Macro(macro_def, macro_kind)
             }
+            ItemKind::Delegation(box delegation) => self.with_new_scopes(ident.span, |this| {
+                this.current_item = Some(ident.span);
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/item.rs:431]: This doesn't seem to be necessary.

File: compiler/rustc_ast_lowering/src/item.rs [hunk 9]
```diff
@@ -787,6 +797,17 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 );
                 (generics, kind, ty.is_some())
             }
+            AssocItemKind::Delegation(box delegation) => {
+                let delegation_results = self.lower_delegation(delegation);
+                (
+                    delegation_results.generics,
+                    hir::TraitItemKind::Fn(
+                        delegation_results.sig,
+                        hir::TraitFn::Provided(delegation_results.body_id),
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/item.rs:806]: Style nit: these multi-line muti-level tuples look horrible.
Please, use separate statements or shorter names to avoid.

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+        let decl = self.lower_delegation_decl(res_id, delegation.span);
+        let sig = self.lower_delegation_sig(delegation.span, decl);
+        let body_id = self.lower_delegation_body(sig.decl, delegation);
+
+        let generics = self.lower_delegation_generics(delegation.span);
+        DelegationResults { body_id, sig, generics }
+    }
+
+    fn get_delegation_res_id(&self, node_id: NodeId) -> Option<DefId> {
+        self.resolver.get_partial_res(node_id).map(|r| r.base_res().opt_def_id()).unwrap_or(None)
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:67]:
```suggestion
        self.resolver.get_partial_res(node_id).map(|r| r.full_res().opt_def_id()).unwrap_or(None)
```
Paths in delegation items must have full resolutions (`PathSource::defer_to_typeck`).

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+    fn lower_delegation_generics(&mut self, span: Span) -> &'hir hir::Generics<'hir> {
+        self.arena.alloc(hir::Generics {
+            params: self.arena.alloc_from_iter([]),
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:72]:
```suggestion
            params: &[],
```
It doesn't make sense to allocate empty slices on arena.

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+            ident: Ident::empty(),
+            hir_id: self.next_id(),
+            res: hir::def::Res::Local(param_id),
+            args: None,
+            infer_args: false,
+        }));
+
+        let path = self.arena.alloc(hir::Path {
+            span: ty.span,
+            res: hir::def::Res::Local(param_id),
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:148]:
```suggestion
            res: Res::Local(param_id),
```
Style nit: imports are preferred to paths with 3+ segments (here and in other places).

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+
+        self.lower_body(|this| {
+            let mut parameters: Vec<hir::Param<'_>> = Vec::new();
+            let mut args: Vec<hir::Expr<'hir>> = Vec::new();
+
+            for (idx, param_ty) in decl.inputs.iter().enumerate() {
+                let (param, pat_node_id) = this.generate_param(param_ty);
+                parameters.push(param);
+
+                if idx != 0 || expr.is_none() {
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:183]:
```suggestion
                if let Some(expr) && idx == 0 {
```
Reversing the condition would be clearer.

Also could you add a test for delegation case with `self` being used in the body, but the first argument not actually existing?

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+                    let pat_hir_id = this.lower_node_id(pat_node_id);
+                    let arg = this.generate_arg(param_ty, pat_hir_id);
+                    args.push(arg);
+                } else {
+                    let expr = expr.unwrap();
+                    let mut self_resolver =
+                        SelfResolver { res_id: pat_node_id, resolver: this.resolver };
+                    self_resolver.visit_expr(expr);
+                    let target_expr = this.lower_expr_mut(expr);
+                    args.push(target_expr);
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:193]: The `args.push(...)` part is common for both branches.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 15]
```diff
@@ -2405,13 +2405,119 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+            self.set_tainted_by_errors(e);
+            return Ty::new_error(self.tcx(), e);
+        };
+        let res_sig = self.tcx().fn_sig(res_id);
+
+        let parent = self.tcx().parent(self.item_def_id());
+        let item_def_kind = self.tcx().def_kind(self.item_def_id());
+        let parent_def_kind = self.tcx().def_kind(parent);
+
+        let res_sig = match (item_def_kind, parent_def_kind) {
```
Comment (petrochenkov) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:2491]:
```suggestion
        let res_sig = match parent_def_kind {
```
`item_def_kind` doesn't affect anything here.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 16]
```diff
@@ -2405,13 +2405,119 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                res_sig.instantiate(self.tcx(), substs).skip_binder()
+            }
+            (DefKind::Fn, _) | (DefKind::AssocFn, _) => {
+                res_sig.instantiate_identity().skip_binder()
+            }
+            _ => span_bug!(span, "ty_from_delegation: couldn't calculate resolution signature"),
+        };
+
+        if let hir::InferDelegationVar::Input(id) = idx {
+            res_sig.inputs()[id]
```
Comment (petrochenkov) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:2505]: Style nit: `if let`s with short arms like this typically look better as `match`es.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 17]
```diff
@@ -2405,13 +2405,119 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+        let parent = self.tcx().parent(self.item_def_id());
+        let item_def_kind = self.tcx().def_kind(self.item_def_id());
+        let parent_def_kind = self.tcx().def_kind(parent);
+
+        let res_sig = match (item_def_kind, parent_def_kind) {
+            (DefKind::AssocFn, DefKind::Impl { of_trait: true }) => {
+                let self_ty = self.tcx().type_of(parent).instantiate_identity();
+                let generic_self_ty = ty::GenericArg::from(self_ty);
+                let substs = self.tcx().mk_args_from_iter(std::iter::once(generic_self_ty));
+                res_sig.instantiate(self.tcx(), substs).skip_binder()
```
Comment (petrochenkov) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:2496]: Signature is taken from the wrong item here.
In `impl Trait for Type { delegate anything::foo; }` the signature is supposed to be taken from `Trait::foo`, not from the callee.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 5 above)
Comment (Bryanskiy) [reply compiler/rustc_resolve/src/late/diagnostics.rs:224]: In fact, this is not necessary. The message will be "cannot find function 'function' in 'module'".

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 6 above)
Comment (Bryanskiy) [reply compiler/rustc_resolve/src/late/diagnostics.rs:450]: Fully-qualified path will be suggested.  And this may be misleading, since we do not yet support type-relative paths.

File: compiler/rustc_ast/src/ast.rs [hunk 18]
```diff
@@ -2978,6 +2978,14 @@ pub struct Fn {
     pub body: Option<P<Block>>,
 }
 
+#[derive(Clone, Encodable, Decodable, Debug)]
+pub struct Delegation {
+    pub id: NodeId,
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_ast/src/ast.rs:3007]: The id is necessary because the path doesn't have its own id.

File: compiler/rustc_resolve/src/build_reduced_graph.rs [hunk 19]
```diff
@@ -788,6 +788,11 @@ impl<'a, 'b, 'tcx> BuildReducedGraphVisitor<'a, 'b, 'tcx> {
             ItemKind::Impl(box Impl { of_trait: Some(..), .. }) => {
                 self.r.trait_impl_items.insert(local_def_id);
             }
+
+            ItemKind::Delegation(..) => {
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_resolve/src/build_reduced_graph.rs]: >(Some simple TypeNS arms could also be merged.)

^

File: compiler/rustc_parse/src/parser/item.rs [hunk 20]
```diff
@@ -255,6 +255,8 @@ impl<'a> Parser<'a> {
         {
             // IMPL ITEM
             self.parse_item_impl(attrs, def_())?
+        } else if self.check_keyword(kw::Reuse) {
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_parse/src/parser/item.rs]: >`reuse::my_macro!();` in item position to parsing tests

^

File: compiler/rustc_resolve/src/late.rs (hunk 3 above)
Comment (petrochenkov) [reply compiler/rustc_resolve/src/late.rs:3189]: >Also, recursive visiting typically happens after the current node visiting, `smart_resolve_path` in this case.

^

File: compiler/rustc_ast/src/ast.rs [hunk 21]
```diff
@@ -3063,14 +3071,20 @@ pub enum ItemKind {
 
     /// A macro definition.
     MacroDef(MacroDef),
+
+    /// A delegation item (`reuse`).
+    ///
+    /// E.g. `reuse <Type as Trait>::name { target_expr_template }`.
+    Delegation(Box<Delegation>),
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_ast/src/ast.rs:3102]: ^ Could you check this?

[the remark this answers is not in the document]
Comment (Bryanskiy) [reply compiler/rustc_ast/src/ast.rs:3102]: It doesn't fit

File: compiler/rustc_ast_lowering/src/item.rs [hunk 22]
```diff
@@ -921,12 +947,16 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
                 }
+                AssocItemKind::Delegation(..) => {
+                    hir::AssocItemKind::Fn { has_self: self.delegation_has_self(i.id, i.span) }
+                }
                 AssocItemKind::MacCall(..) => unimplemented!(),
             },
             trait_item_def_id: self
                 .resolver
                 .get_partial_res(i.id)
-                .map(|r| r.expect_full_res().def_id()),
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/item.rs:929]: Why this change?
We should never put partial resolution into the table for these IDs due to `defer_to_typeck == false` for `PathSource::Delegation`.

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 23]
```diff
@@ -0,0 +1,306 @@
...
+        let generics = self.lower_delegation_generics(span);
+        DelegationResults { body_id, sig, generics }
+    }
+
+    fn get_delegation_res_id(&self, node_id: NodeId, span: Span) -> Option<DefId> {
+        let res = self
+            .resolver
+            .get_partial_res(node_id)
+            .map(|or| or.full_res().map(|r| r.opt_def_id()).unwrap_or(None))
+            .unwrap_or(None);
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:67]: Resolver always sets resolutions for these ids, even in case of errors, no?
And they are never partial due to `defer_to_typeck == false` for `PathSource::Delegation`.
We should be able to get rid of all these `Option`s.

```rust
let res = self.resolver.get_partial_res(node_id).unwrap().expect_full_res().opt_def_id();
```

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 24]
```diff
@@ -0,0 +1,306 @@
...
+    fn visit_path(&mut self, path: &'ast Path, id: NodeId) {
+        self.try_replace_id(id);
+        visit::walk_path(self, path);
+    }
+
+    fn visit_path_segment(&mut self, path_segment: &'ast PathSegment) {
+        if kw::SelfLower == path_segment.ident.name {
+            self.try_replace_id(path_segment.id);
+        }
+    }
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:305]: Missing recursive `walk_path_segment`.

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 25]
```diff
@@ -0,0 +1,306 @@
...
+}
+
+impl<'ast, 'a> Visitor<'ast> for SelfResolver<'a> {
+    fn visit_path(&mut self, path: &'ast Path, id: NodeId) {
+        self.try_replace_id(id);
+        visit::walk_path(self, path);
+    }
+
+    fn visit_path_segment(&mut self, path_segment: &'ast PathSegment) {
+        if kw::SelfLower == path_segment.ident.name {
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:302]: It's unlikely that this optimization is useful.

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 26]
```diff
@@ -0,0 +1,306 @@
...
+    fn lower_delegation_body(
+        &mut self,
+        decl: &'hir hir::FnDecl<'hir>,
+        delegation: &Delegation,
+        node_id: NodeId,
+        span: Span,
+    ) -> BodyId {
+        let path_id = if self.is_in_trait_impl { delegation.id } else { node_id };
+        let path = self.lower_qpath(
+            path_id,
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:175]:
```suggestion
            delegation.id,
```
Isn't the path always resolved into `delegation.id`?
That's the whole point of that ID.

File: compiler/rustc_resolve/src/late.rs [hunk 27]
```diff
@@ -3148,6 +3178,36 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
...
+    ) {
+        let (qself, path) = (&delegation.qself, &delegation.path);
+        if let Some(qself) = &delegation.qself {
+            self.visit_ty(&qself.ty);
+        }
+        self.visit_path(path, delegation.id);
+        self.smart_resolve_path(delegation.id, qself, path, PathSource::Delegation);
+        if !is_trait_impl {
+            let path_res = self.r.partial_res_map.get(&delegation.id).unwrap();
+            self.r.partial_res_map.insert(item_id, *path_res);
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_resolve/src/late.rs:3191]: But the main question, I guess, is why this is required.
We should never retrieve the path's resolution from item's id, we have `delegation.id` for that.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 28]
```diff
@@ -2331,13 +2331,114 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+            error_occured = true;
+        };
+
+        if let Some(node) = self.tcx().hir().get_if_local(res_id)
+            && let Some(decl) = node.fn_decl()
+            && let hir::FnRetTy::Return(ty) = decl.output
+        {
+            match ty.kind {
+                hir::TyKind::InferDelegation(_, _) => try_emit("recursive delegation"),
+                hir::TyKind::OpaqueDef(_, _, _) => {
```
Comment (petrochenkov) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs]: Opaque type may also be inside the return type, like `Option<impl Trait>`.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 29]
```diff
@@ -2331,13 +2331,114 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+        }
+
+        if self.tcx().constness(res_id) == hir::Constness::Const {
+            try_emit("delegation to const functions");
+        }
+
+        let res_sig = self.tcx().fn_sig(res_id).instantiate_identity();
+        if res_sig.c_variadic() {
+            try_emit("delegation to variadic functions");
+            return error_occured;
```
Comment (petrochenkov) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:2406]: `error_occured` is returned here, but not in other checks, not sure why.

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 30]
```diff
@@ -0,0 +1,306 @@
...
+                DefKind::AssocFn => self.tcx.associated_item(res_id).fn_has_self_parameter,
+                _ => span_bug!(span, "unexpected DefKind for delegation item"),
+            }
+        }
+    }
+
+    pub fn lower_delegation(
+        &mut self,
+        delegation: &Delegation,
+        node_id: NodeId,
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/delegation.rs:46]: It is not clear what these `node_id`s passed around mean, the naming is very generic.
It is also not clear where the selection between the `resolution_of(item_id)` for trait impl items, and `resolution_of(delegation.id)` for all other items happens.
Please make it clear and commented.
(The selection should fully happen here in AST lowering, and not in resolve.)

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 31]
```diff
@@ -590,6 +590,10 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
         res: Option<Res>,
         base_error: &BaseError,
     ) -> (bool, Vec<ImportSuggestion>) {
+        if let PathSource::Delegation = source {
+            return (false, vec![]);
+        }
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:595]: https://github.com/petrochenkov/rust/commit/63ee3d6b9746639494630e58d6b416f6431d08ce

File: compiler/rustc_ast_lowering/src/delegation.rs (hunk 30 above)
Comment (petrochenkov) [reply compiler/rustc_ast_lowering/src/delegation.rs:46]: I suggest using some consistent naming scheme to improve this, e.g.
- `item_id` for ids of delegation items
- `path_id` for ids of delegation paths (may be unnecessary because the is always `delegation.id`)
- `sig_id` for ids of items from which the signature is inherited

File: compiler/rustc_ast_lowering/src/delegation.rs (hunk 23 above)
Comment (Bryanskiy) [reply compiler/rustc_ast_lowering/src/delegation.rs:67]: No. `check_trait_item` will not set the item id resolution in case of errors.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs (hunk 29 above)
Comment (Bryanskiy) [reply compiler/rustc_hir_analysis/src/astconv/mod.rs:2407]: Comment was added.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 32]
```diff
@@ -2332,13 +2333,134 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+        if let Some(node) = self.tcx().hir().get_if_local(res_id)
+            && let Some(decl) = node.fn_decl()
+            && let hir::FnRetTy::Return(ty) = decl.output
+            && let hir::TyKind::InferDelegation(_, _) = ty.kind
+        {
+            try_emit("recursive delegation");
+        }
+
+        let res_sig = self.tcx().fn_sig(res_id).instantiate_identity();
+        if self.check_delegation_for_opaque_types(res_sig) {
```
Comment (petrochenkov) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:2383]:
```suggestion
        if res_sig.output().has_opaque_types() {
```
Something like this should be enough, I think.
(Types in function parameters cannot define opaque types, and using opaque types defined elsewhere should be fine.)

**Comment by petrochenkov** [comment]
The delegation syntax also needs feature gating (via `gated_spans`).
@rustbot author

File: tests/pretty/delegation.rs [hunk 33]
```diff
@@ -0,0 +1,25 @@
+#![feature(fn_delegation)]
+//~^ WARN the feature `fn_delegation` is incomplete
```
Comment (petrochenkov) [inline tests/pretty/delegation.rs:2]: Probably better allow `incomplete_features` in all these tests.

**Comment by petrochenkov** [comment]
r? @lcnr 
Could you check the astconv part?
The general idea is that some features are not yet supported (see `check_delegation_constraints` for the list), but the supported ones should behave correctly and not ICE.

**Review by petrochenkov — APPROVED** [review]

File: src/tools/rustfmt/src/visitor.rs [hunk 34]
```diff
@@ -592,6 +592,7 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
                     );
                     self.push_rewrite(item.span, rewrite);
                 }
+                ast::ItemKind::Delegation(..) => {}
```
Comment (ytmimi) [inline src/tools/rustfmt/src/visitor.rs:595]: Might want to change this so rustfmt doesn't totally remove the Delegation item.

```suggestion
                ast::ItemKind::Delegation(..) => {
                    // TODO: rewrite delegation items once syntax is established.
                    // For now, leave the contents of the Span unformatted.
                    self.push_rewrite(item.span, None)
                }
```

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 35]
```diff
@@ -0,0 +1,317 @@
...
+use rustc_middle::ty::ResolverAstLowering;
+use rustc_span::{symbol::Ident, Span};
+use rustc_target::spec::abi;
+use std::iter;
+
+// id's naming scheme:
+//
+// item_id - id of delegation item
+// path_id - id of delegation path
+// sig_id - id of item from which the signature is inherited
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/delegation.rs:23]: Could you give a sketch of what the ast->hir transformation looks like? This file appears to do a lot of non-trivial constructions that could use a summarised form.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 36]
```diff
@@ -2332,13 +2331,115 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+        let sig = self.tcx().fn_sig(sig_id).instantiate_identity();
+        if sig.output().has_opaque_types() {
+            try_emit("delegation to a function with opaque type");
+        }
+
+        let res_generics = self.tcx().generics_of(sig_id);
+        let parent = self.tcx().parent(self.item_def_id());
+        let parent_generics = self.tcx().generics_of(parent);
+        let self_ty_bias = (self.tcx().def_kind(parent) == DefKind::Trait) as usize;
+        if parent_generics.count() + generics_param_count(res_generics.own_counts()) > self_ty_bias
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:2370]: this is simply `res_generics.count()`, is it not?

File: compiler/rustc_ast_lowering/src/delegation.rs (hunk 35 above)
Comment (lcnr) [reply compiler/rustc_ast_lowering/src/delegation.rs:51]: please do provide an example desugaring here

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 37]
```diff
@@ -2332,13 +2331,115 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+        let parent_def_kind = self.tcx().def_kind(parent);
+
+        let sig = if let DefKind::Impl { of_trait: true } = parent_def_kind {
+            let self_ty = self.tcx().type_of(parent).instantiate_identity();
+            let generic_self_ty = ty::GenericArg::from(self_ty);
+            let substs = self.tcx().mk_args_from_iter(std::iter::once(generic_self_ty));
+            sig.instantiate(self.tcx(), substs).skip_binder()
+        } else {
+            sig.instantiate_identity().skip_binder()
+        };
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:2425]: this relies on there being no other generic params except the trait self type, please add this as a comment. And potentially as an assert

Comment (lcnr) [reply compiler/rustc_hir_analysis/src/astconv/mod.rs:2425]: Please also add a comment explaining why `skip_binder` is acceptable here and where we rebind it, given that the delegated to function can have late bound params.

**Review by lcnr — COMMENTED** [review]
If I understand correctly, you implement delegation by lowering
```rust
struct A;
struct B;
trait Trait {
    fn foo(&self, x: u32) -> i32;
}
impl Trait for A {
    fn foo(&self, _: u32) -> i32 { 1 }
}
impl Trait for B {
    reuse <A as Trait>::foo { &A }
}
```
to
```rust
// ...

impl Trait for B {
    fn foo(InferDelegation(<A as Trait>::foo, Input(0)), InferDelegation(<A as Trait>::foo, Input(1))) -> InferDelegation(<A as Trait>::foo, Output)
}
```

I personally feel somewhat unhappy with the addition of `InferDelegation` but can see why it was chosen and can't immediately think of a better alternative

File: compiler/rustc_hir_analysis/src/astconv/mod.rs (hunk 36 above)
Comment (Bryanskiy) [reply compiler/rustc_hir_analysis/src/astconv/mod.rs:2370]: Consider the example
```
trait Trait<T0> {
    fn foo<T1>() {}
}

struct Struct<T2> { /* ... */ }

impl<T0, T2> Trait<T0> for Struct<T2>  {
    /* reuse Trait::foo */
}
```
For a delegation item in a trait impl `sig_id` is a trait method resolution. So, `res_generics = [Self, T1]`, `parent_generics = [T0, T2]`. Expected generics for the `foo`: `[T0, T1, T2]`. That's where the addition comes from.

**Comment by petrochenkov** [comment]
r=me after rebase.

**Comment by lcnr** [comment]
still open astconv issues, going to reassign to myself for final review there

r? @lcnr

**Comment by lcnr** [comment]
another thought: afaict you only use the `hir::TyKind::DelegationInfer` as part of the `hir::FnDecl`. I think it may be preferable to instead change `FnDecl` to an enum with either a proper `FnDecl` or a `DelegationDummy`. This way less code has to think about how to potentially handle delegation

File: compiler/rustc_hir_analysis/src/astconv/mod.rs (hunk 36 above)
Comment (Bryanskiy) [reply compiler/rustc_hir_analysis/src/astconv/mod.rs]: My mistake. Conditions have been replaced.

**Comment by Bryanskiy** [comment]
> another thought: afaict you only use the `hir::TyKind::DelegationInfer` as part of the `hir::FnDecl`. I think it may be preferable to instead change `FnDecl` to an enum with either a proper `FnDecl` or a `DelegationDummy`. This way less code has to think about how to potentially handle delegation

The change looks very large. I think it should be in another PR. @petrochenkov?

**Comment by petrochenkov** [comment]
Yeah, I expect that to be a quite invasive change.
The benefit of the current representation is that delegation almost entirely disappears during AST lowering - if you need to access the function's HIR later it would look like a regular function without special cases (with exception of astconv).

We may need something like this for [step 2](https://github.com/rust-lang/rfcs/blob/923b8e65639de80a93ef2b9770ba1c976c24cf5d/text/0000-fn-delegation.md#implementation-steps) (type-relative paths in delegation) because with those we wouldn't have access to signatures (to parameter counts in particular) during AST lowering.
But I'd prefer completing step 1 + multiple name support before doing anything related to step 2.

**Comment by petrochenkov** [comment]
tidy needs fixing.
@rustbot author
